File: lambda-deploy/shared/python/middleware/examples/mfa_handler.py
"""
Example Lambda handler demonstrating MFA integration
This shows how to handle MFA challenges in authentication flows
"""

import os
import json
import boto3
from botocore.exceptions import ClientError
from ..auth import create_auth_middleware


def protected_handler(event, context):
    """
    Example: Protected endpoint requiring MFA-authenticated users
    """
    auth_middleware = create_auth_middleware()
    auth_result = auth_middleware(event)

    if not auth_result['authorized']:
        return {
            'statusCode': 401,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'Unauthorized', 'message': auth_result['error']}),
        }

    # Access user information
    user = event['user']

    # Check if user has MFA enabled (optional additional check)
    has_mfa = user['claims'].get('cognito:mfa_enabled') == 'true'

    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
        },
        'body': json.dumps(
            {
                'message': 'Access granted',
                'user': {
                    'username': user['username'],
                    'email': user['email'],
                    'mfaEnabled': has_mfa,
                },
            }
        ),
    }


def mfa_status_handler(event, context):
    """
    Example: Endpoint to check MFA status
    """
    auth_middleware = create_auth_middleware()
    auth_result = auth_middleware(event)

    if not auth_result['authorized']:
        return {
            'statusCode': 401,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'Unauthorized', 'message': auth_result['error']}),
        }

    user = event['user']

    # Extract MFA information from Cognito claims
    mfa_enabled = user['claims'].get('cognito:mfa_enabled') == 'true'
    preferred_mfa = user['claims'].get('cognito:preferred_mfa', 'NONE')

    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
        },
        'body': json.dumps(
            {'mfaEnabled': mfa_enabled, 'preferredMfa': preferred_mfa, 'username': user['username']}
        ),
    }


def custom_auth_handler(event, context):
    """
    Example: Custom authentication flow with MFA handling
    This demonstrates how to handle MFA challenges in a custom auth flow
    """
    client = boto3.client('cognito-idp', region_name=os.environ.get('AWS_REGION', 'ap-south-1'))

    try:
        body = json.loads(event.get('body', '{}'))
        username = body.get('username')
        password = body.get('password')
        mfa_code = body.get('mfaCode')
        session = body.get('session')
        challenge_name = body.get('challengeName')

        # If MFA code is provided, respond to MFA challenge
        if mfa_code and session and challenge_name:
            response = client.respond_to_auth_challenge(
                ClientId=os.environ['COGNITO_CLIENT_ID'],
                ChallengeName=challenge_name,  # 'SMS_MFA' or 'SOFTWARE_TOKEN_MFA'
                Session=session,
                ChallengeResponses={
                    'USERNAME': username,
                    (
                        'SMS_MFA_CODE' if challenge_name == 'SMS_MFA' else 'SOFTWARE_TOKEN_MFA_CODE'
                    ): mfa_code,
                },
            )

            return {
                'statusCode': 200,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                },
                'body': json.dumps(
                    {'success': True, 'authenticationResult': response.get('AuthenticationResult')}
                ),
            }

        # Initial authentication
        response = client.initiate_auth(
            AuthFlow='USER_PASSWORD_AUTH',
            ClientId=os.environ['COGNITO_CLIENT_ID'],
            AuthParameters={'USERNAME': username, 'PASSWORD': password},
        )

        # Check if MFA challenge is required
        if response.get('ChallengeName') in ['SMS_MFA', 'SOFTWARE_TOKEN_MFA']:
            return {
                'statusCode': 200,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                },
                'body': json.dumps(
                    {
                        'success': False,
                        'mfaRequired': True,
                        'challengeName': response['ChallengeName'],
                        'session': response['Session'],
                        'message': (
                            'MFA code sent to your phone'
                            if response['ChallengeName'] == 'SMS_MFA'
                            else 'Enter code from your authenticator app'
                        ),
                    }
                ),
            }

        # Authentication successful without MFA
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps(
                {'success': True, 'authenticationResult': response.get('AuthenticationResult')}
            ),
        }

    except ClientError as error:
        print(f'Authentication error: {error}')

        return {
            'statusCode': 400,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'Authentication failed', 'message': str(error)}),
        }


def sensitive_operation_handler(event, context):
    """
    Example: Require MFA for sensitive operations
    """
    auth_middleware = create_auth_middleware()
    auth_result = auth_middleware(event)

    if not auth_result['authorized']:
        return {
            'statusCode': 401,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'Unauthorized', 'message': auth_result['error']}),
        }

    user = event['user']
    mfa_enabled = user['claims'].get('cognito:mfa_enabled') == 'true'

    # Require MFA for sensitive operations
    if not mfa_enabled:
        return {
            'statusCode': 403,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps(
                {
                    'error': 'MFA Required',
                    'message': 'This operation requires multi-factor authentication to be enabled',
                }
            ),
        }

    # Perform sensitive operation
    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
        },
        'body': json.dumps({'message': 'Sensitive operation completed successfully'}),
    }

File: backend/shared/nodejs/middleware/examples/mfa-handler.js
/**
 * Example Lambda handler demonstrating MFA integration
 * This shows how to handle MFA challenges in authentication flows
 */

const { createAuthMiddleware } = require('../auth');

/**
 * Example: Protected endpoint requiring MFA-authenticated users
 */
exports.protectedHandler = async (event, context) => {
  // Apply authentication middleware
  const authMiddleware = createAuthMiddleware();
  const authResult = await authMiddleware(event);

  if (!authResult.authorized) {
    return {
      statusCode: 401,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        error: 'Unauthorized',
        message: authResult.error,
      }),
    };
  }

  // Access user information
  const user = event.user;

  // Check if user has MFA enabled (optional additional check)
  const hasMFA = user.claims['cognito:mfa_enabled'] === 'true';

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
    },
    body: JSON.stringify({
      message: 'Access granted',
      user: {
        username: user.username,
        email: user.email,
        mfaEnabled: hasMFA,
      },
    }),
  };
};

/**
 * Example: Endpoint to check MFA status
 */
exports.mfaStatusHandler = async (event, context) => {
  const authMiddleware = createAuthMiddleware();
  const authResult = await authMiddleware(event);

  if (!authResult.authorized) {
    return {
      statusCode: 401,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        error: 'Unauthorized',
        message: authResult.error,
      }),
    };
  }

  const user = event.user;

  // Extract MFA information from Cognito claims
  const mfaEnabled = user.claims['cognito:mfa_enabled'] === 'true';
  const preferredMfa = user.claims['cognito:preferred_mfa'];

  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
    },
    body: JSON.stringify({
      mfaEnabled,
      preferredMfa: preferredMfa || 'NONE',
      username: user.username,
    }),
  };
};

/**
 * Example: Custom authentication flow with MFA handling
 * This demonstrates how to handle MFA challenges in a custom auth flow
 */
exports.customAuthHandler = async (event, context) => {
  const {
    CognitoIdentityProviderClient,
    InitiateAuthCommand,
    RespondToAuthChallengeCommand,
  } = require('@aws-sdk/client-cognito-identity-provider');

  const client = new CognitoIdentityProviderClient({
    region: process.env.AWS_REGION || 'ap-south-1',
  });

  try {
    const body = JSON.parse(event.body || '{}');
    const { username, password, mfaCode, session, challengeName } = body;

    // If MFA code is provided, respond to MFA challenge
    if (mfaCode && session && challengeName) {
      const command = new RespondToAuthChallengeCommand({
        ClientId: process.env.COGNITO_CLIENT_ID,
        ChallengeName: challengeName, // 'SMS_MFA' or 'SOFTWARE_TOKEN_MFA'
        Session: session,
        ChallengeResponses: {
          USERNAME: username,
          [challengeName === 'SMS_MFA' ? 'SMS_MFA_CODE' : 'SOFTWARE_TOKEN_MFA_CODE']: mfaCode,
        },
      });

      const response = await client.send(command);

      return {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json',
          'Access-Control-Allow-Origin': '*',
        },
        body: JSON.stringify({
          success: true,
          authenticationResult: response.AuthenticationResult,
        }),
      };
    }

    // Initial authentication
    const command = new InitiateAuthCommand({
      AuthFlow: 'USER_PASSWORD_AUTH',
      ClientId: process.env.COGNITO_CLIENT_ID,
      AuthParameters: {
        USERNAME: username,
        PASSWORD: password,
      },
    });

    const response = await client.send(command);

    // Check if MFA challenge is required
    if (response.ChallengeName === 'SMS_MFA' || response.ChallengeName === 'SOFTWARE_TOKEN_MFA') {
      return {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json',
          'Access-Control-Allow-Origin': '*',
        },
        body: JSON.stringify({
          success: false,
          mfaRequired: true,
          challengeName: response.ChallengeName,
          session: response.Session,
          message:
            response.ChallengeName === 'SMS_MFA'
              ? 'MFA code sent to your phone'
              : 'Enter code from your authenticator app',
        }),
      };
    }

    // Authentication successful without MFA
    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        success: true,
        authenticationResult: response.AuthenticationResult,
      }),
    };
  } catch (error) {
    console.error('Authentication error:', error);

    return {
      statusCode: 400,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        error: 'Authentication failed',
        message: error.message,
      }),
    };
  }
};

/**
 * Example: Require MFA for sensitive operations
 */
exports.sensitiveOperationHandler = async (event, context) => {
  const authMiddleware = createAuthMiddleware();
  const authResult = await authMiddleware(event);

  if (!authResult.authorized) {
    return {
      statusCode: 401,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        error: 'Unauthorized',
        message: authResult.error,
      }),
    };
  }

  const user = event.user;
  const mfaEnabled = user.claims['cognito:mfa_enabled'] === 'true';

  // Require MFA for sensitive operations
  if (!mfaEnabled) {
    return {
      statusCode: 403,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify({
        error: 'MFA Required',
        message: 'This operation requires multi-factor authentication to be enabled',
      }),
    };
  }

  // Perform sensitive operation
  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
    },
    body: JSON.stringify({
      message: 'Sensitive operation completed successfully',
    }),
  };
};
